src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp

```diff
@@ -261,6 +264,90 @@ ov::npuw::LLMInferRequest::LLMInferRequest(const std::shared_ptr<ov::npuw::LLMCo
     m_gemma_sliding_window_size = compiled_model->m_gemma_sliding_window_size;
 }
 
+void ov::npuw::LLMInferRequest::create_generate_request_variants(
+    const std::shared_ptr<ov::npuw::LLMCompiledModel>& compiled_model) {
+    // Create multiple KV cache model variant requests
+    m_kvcache_sizes = compiled_model->m_kvcache_sizes;
+    m_generate_requests.reserve(compiled_model->m_generate_compiled_variants.size());
+
+    // First, create the largest variant request (last one in the list)
+    auto largest_generate_request = compiled_model->m_generate_compiled_variants.back()->create_infer_request();
+
+    // Store past KV tensors from the largest variant for sharing
+    std::unordered_map<std::string, ov::SoPtr<ov::ITensor>> largest_past_kv_tensors;
+    for (const auto& input_port : largest_generate_request->get_compiled_model()->inputs()) {
+        const auto& input_name = input_port.get_any_name();
+        if (input_name.find(layer_names::past_key_values) != std::string::npos) {
+            largest_past_kv_tensors[input_name] = largest_generate_request->get_tensor(input_port);
+        }
+    }
+
+    // Create all variant requests and share past KV tensors
+    for (size_t i = 0; i < compiled_model->m_generate_compiled_variants.size(); ++i) {
+        std::shared_ptr<ov::IAsyncInferRequest> generate_request;
+
+        if (i == compiled_model->m_generate_compiled_variants.size() - 1) {
+            // Use the already created largest variant
+            generate_request = largest_generate_request;
+        } else {
+            // Create smaller variant
+            generate_request = compiled_model->m_generate_compiled_variants[i]->create_infer_request();
+
+            // Share past KV tensors from the largest variant
+            for (const auto& input_port : generate_request->get_compiled_model()->inputs()) {
+                const auto& input_name = input_port.get_any_name();
+                if (input_name.find(layer_names::past_key_values) != std::string::npos) {
+                    if (largest_past_kv_tensors.find(input_name) != largest_past_kv_tensors.end()) {
```

Comment:
Can we have an `else` here or there should be an `assert` that we can't?
We might use `largest_past_kv_tensors.at(input_name)` instead